File: internal/models/schedule.go
package models

import "time"

// ScheduleEventType — кесте оқиға түрі
type ScheduleEventType string

const (
	ScheduleLesson   ScheduleEventType = "lesson"
	ScheduleDeadline ScheduleEventType = "deadline"
	ScheduleExam     ScheduleEventType = "exam"
	ScheduleHoliday  ScheduleEventType = "holiday"
	ScheduleMeeting  ScheduleEventType = "meeting"
	ScheduleOther    ScheduleEventType = "other"
)

type ScheduleEvent struct {
	ID           uint           `gorm:"primaryKey"                     json:"id"`
	TeacherID    uint           `gorm:"not null;index"                 json:"teacher_id"`
	CourseID     *uint          `gorm:"index"                          json:"course_id,omitempty"`
	AssignmentID *uint          `gorm:"index"                          json:"assignment_id,omitempty"`
	Title        string         `gorm:"size:255;not null"              json:"title"`
	Description  string         `gorm:"type:text;default:''"           json:"description"`
	StartTime    time.Time      `gorm:"not null"                       json:"start"`
	EndTime      *time.Time     `                                      json:"end,omitempty"`
	AllDay       bool           `gorm:"default:false"                  json:"all_day"`
	Type         ScheduleEventType `gorm:"size:50;default:'lesson'"     json:"type"`
	Color        string         `gorm:"size:20;default:''"             json:"color"`
	CreatedAt    time.Time      `                                      json:"created_at"`
	UpdatedAt    time.Time      `                                      json:"updated_at"`
}

type CreateScheduleEventRequest struct {
	CourseID     *uint             `json:"course_id"`
	AssignmentID *uint             `json:"assignment_id"`
	Title        string            `json:"title" binding:"required"`
	Description  string            `json:"description"`
	Start        string            `json:"start" binding:"required"`
	End          string            `json:"end"`
	AllDay       bool              `json:"all_day"`
	Type         ScheduleEventType `json:"type"`
	Color        string            `json:"color"`
}

type UpdateScheduleEventRequest struct {
	Title       *string            `json:"title"`
	Description *string            `json:"description"`
	Start       *string            `json:"start"`
	End         *string            `json:"end"`
	AllDay      *bool              `json:"all_day"`
	Type        *ScheduleEventType `json:"type"`
	Color       *string            `json:"color"`
	CourseID    *uint              `json:"course_id"`
}

File: lib/api/schedule.ts
import { apiClient } from "@/lib/auth/api-client"

export type EventType = "lesson" | "deadline" | "exam" | "holiday" | "meeting" | "other"

export interface ScheduleEvent {
  id: number
  teacher_id: number
  course_id?: number
  assignment_id?: number
  title: string
  description: string
  start: string       // ISO 8601
  end?: string        // ISO 8601
  all_day: boolean
  type: EventType
  color: string
  created_at: string
  updated_at: string
}

export interface CreateEventRequest {
  title: string
  description?: string
  start: string
  end?: string
  all_day?: boolean
  type?: EventType
  color?: string
  course_id?: number
  assignment_id?: number
}

export interface UpdateEventRequest {
  title?: string
  description?: string
  start?: string
  end?: string
  all_day?: boolean
  type?: EventType
  color?: string
  course_id?: number
}

export async function listEvents(params?: {
  start?: string
  end?: string
  courseId?: number
}): Promise<ScheduleEvent[]> {
  const qs = new URLSearchParams()
  if (params?.start)    qs.set("start", params.start)
  if (params?.end)      qs.set("end", params.end)
  if (params?.courseId) qs.set("course_id", String(params.courseId))
  const q = qs.toString()
  return apiClient.get(`/api/teacher/schedule${q ? `?${q}` : ""}`)
}

export async function createEvent(data: CreateEventRequest): Promise<ScheduleEvent> {
  return apiClient.post("/api/teacher/schedule", data)
}

export async function updateEvent(id: number, data: UpdateEventRequest): Promise<ScheduleEvent> {
  return apiClient.put(`/api/teacher/schedule/${id}`, data)
}

export async function deleteEvent(id: number): Promise<void> {
  return apiClient.delete(`/api/teacher/schedule/${id}`)
}

export async function syncDeadlines(): Promise<{ created: number }> {
  return apiClient.post("/api/teacher/schedule/sync", {})
}

// ── FullCalendar event color palette ────────────────────────────────────────

export const EVENT_COLORS: Record<EventType, string> = {
  lesson:   "#6366f1",
  deadline: "#ef4444",
  exam:     "#f97316",
  holiday:  "#22c55e",
  meeting:  "#3b82f6",
  other:    "#8b5cf6",
}

export const EVENT_LABELS: Record<EventType, string> = {
  lesson:   "Сабақ",
  deadline: "Deadline",
  exam:     "Емтихан",
  holiday:  "Демалыс",
  meeting:  "Кездесу",
  other:    "Басқа",
}
